crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,585 @@
+            OccupiedEntry::Map(e) => e.get_mut(),
+        }
+    }
+
+    /// see [HashMap::OccupiedEntry::into_mut](https://doc.rust-lang.org/std/collections/hash_map/enum.OccupiedEntry.html#method.into_mut)
+    pub fn into_mut(self) -> &'a mut V {
+        match self {
+            OccupiedEntry::List { list, index } => &mut list[index].1,
+            OccupiedEntry::Map(e) => e.into_mut(),
+        }
+    }
+
+    /// see [HashMap::OccupiedEntry::remove](https://doc.rust-lang.org/std/collections/hash_map/enum.OccupiedEntry.html#method.remove)
+    pub fn remove(self) -> V {
+        match self {
+            OccupiedEntry::List { list, index } => list.swap_remove(index).1,
+            OccupiedEntry::Map(e) => e.remove(),
+        }
+    }
+}
+
+pub enum VacantEntry<'a, K, V> {
+    List { list: &'a mut Vec<(K, V)>, key: K },
+    Map(std::collections::hash_map::VacantEntry<'a, K, V>),
+}
+
+impl<'a, K, V> VacantEntry<'a, K, V> {
+    /// see [HashMap::VacantEntry::insert](https://doc.rust-lang.org/std/collections/hash_map/enum.VacantEntry.html#method.insert)
+    pub fn insert(self, value: V) -> &'a mut V {
+        match self {
+            VacantEntry::List { list, key } => {
+                list.push((key, value));
+                &mut list.last_mut().unwrap().1
+            }
+            VacantEntry::Map(entry) => entry.insert(value),
+        }
+    }
+}
+
+impl<K, V, H> Serialize for AutoMap<K, V, H>
+where
+    K: Eq + Hash + Serialize,
+    V: Serialize,
+    H: BuildHasher,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            AutoMap::List(list) => {
+                let mut map = serializer.serialize_map(Some(list.len()))?;
+                for (k, v) in list {
+                    map.serialize_entry(k, v)?;
+                }
+                map.end()
+            }
+            AutoMap::Map(map) => (**map).serialize(serializer),
+        }
+    }
+}
+
+impl<'de, K, V, H> Deserialize<'de> for AutoMap<K, V, H>
+where
+    K: Eq + Hash + Deserialize<'de>,
+    V: Deserialize<'de>,
+    H: BuildHasher + Default,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct AutoMapVisitor<K, V, H> {
+            phantom: PhantomData<AutoMap<K, V, H>>,
+        }
+
+        impl<'de, K, V, H> Visitor<'de> for AutoMapVisitor<K, V, H>
+        where
+            K: Eq + Hash + Deserialize<'de>,
+            V: Deserialize<'de>,
+            H: BuildHasher + Default,
+        {
+            type Value = AutoMap<K, V, H>;
+
+            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+                formatter.write_str("a map")
+            }
+
+            fn visit_map<M>(self, mut m: M) -> Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'de>,
+            {
+                if let Some(size) = m.size_hint() {
```

Comment:
Is the size hint guaranteed? Should we add checks to convert to the appropriate type if under/oversized?

Reply:
I read it like it's guaranteed if provided.